Backports v0.22.3

haampie opened the pull request:
- Set version to v0.22.3.dev0
- [x] #46775
- [x] #46983
- [x] #46966
- [x] #46503 (partially: the bit that bumps archspec)
- [x] #46958
- [x] #46711
- [x] #47165
- [x] docs search improvements: #47107, #47110, #47111
- [x] #47105
- [x] #47035 
- [x] #44374
- [x] #44917
- [x] #45714
- [x] #46980
- [x] #46991
- [x] #47102
- [x] #46709
- [x] #47080
- [x] #47175
- [x] #47197
- [x] #47278
- [x] #47307
- [x] #47308
- [x] #47309
- [x] #47328
- [x] #47327
- [x] #47346
- [x] #47402
- [x] #47358
- [x] #44995
- [x] #47351
- [x] #47434

haampie committed:
installer: improve init signature and explicits (#44374)

Change the installer to take `([pkg], args)` in the constructor instead
of `[(pkg, args)]`. The reason is that certain arguments are global
settings, and the new API ensures that those arguments cannot be
different across different "build requests".

The `explicit` install arg is now a list of hashes, and the installer is
no longer responsible for determining what package is installed
explicitly. This way environment installs can simply pass the list of
environment roots, without them necessarily being explicit build
requests. For example an env with two roots [a, b], where b depends on
a, would not always cause spack install to mark b as explicit.

Notice that `overwrite` already took a list of hashes, this makes
`explicit` consistent.

`package.do_install(explicit=True)` continues to take a boolean.

haampie committed:
installer.py: handle external roots the same (#44917)

There was logic not to enqueue build requests for externals if they occur as
roots. That's unnecessary.

haampie committed:
spack_yaml: add anchorify function (#44995)

This adds spack.util.spack_yaml.anchorify, which takes a non-cyclic
dict/list structure, and replaces identical values with (back)
references to the first instance, so that yaml serialization will use
anchors.

`repr` is used to identify sub-dags, which in principle is quadratic
complexity in depth of the graph, but in practice the depth is O(1) so
this should not matter.

Then this is used in CI to reduce the size of generated YAML files to
30% of their original size.

wdconinc committed:
ffmpeg: update patch hashes for addition of the X-Git-Tag (#45574)

haampie committed:
Show underlying errors on fetch failure (#45714)

- unwrap/flatten nested exceptions
- improve tests
- unify curl lookup

alalazo committed:
Bump archspec 0.2.5-dev (#46503)

Use commit bceb39528ac49dd0c876b2e9bf3e7482e9c2be4a

haampie committed:
tests: fix wrong install name (#46711)

tgamblin committed:
Better docs and typing for `_setup_pkg_and_run` (#46709)

There was a bit of mystery surrounding the arguments for `_setup_pkg_and_run`. It passes
two file descriptors for handling the `gmake`'s job server in child processes, but they are
unsed in the method.

It turns out that there are good reasons to do this -- depending on the multiprocessing
backend, these file descriptors may be closed in the child if they're not passed
directly to it.

- [x] Document all args to `_setup_pkg_and_run`.
- [x] Document all arguments to `_setup_pkg_and_run`.
- [x] Add type hints for `_setup_pkg_and_run`.
- [x] Refactor exception handling in `_setup_pkg_and_run` so it's easier to add type
      hints. `exc_info()` was problematic because it *can* return `None` (just not
      in the context where it's used).  `mypy` was too dumb to notice this.

Signed-off-by: Todd Gamblin <tgamblin@llnl.gov>

haampie committed:
clingo: fix build with Python 3.13+ (#46775)

Python 3.9 deprecated `PyEval_InitThreads` and kept it as a no-op, and
Python 3.13 deleted the function.

The patch removes calls to it.

alalazo committed:
archspec: update to v0.2.5 (#46958)

Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

haampie committed:
python: drop build-tools tag (#46980)

Remove the `build-tools` tag of python, otherwise these types of
concretizations are possible:

```
py-root
  ^py-pip
    ^python@3.12
  ^python@3.13
```

So, a package would be configured with py-pip using python 3.12, but
installed for 3.13, which does not work.

haampie committed:
support python 3.13 bootstrapping from sources (#46983)

haampie committed:
Set version to v0.22.3.dev0

upsj committed:
env depfile: generate Makefile with absolute script path (#46966)

Co-authored-by: Harmen Stoppels <harmenstoppels@gmail.com>

haampie committed:
Update release documentation (#46991)

haampie committed:
fix use of traceback.format_exception (#47080)

Co-authored-by: Peter Scheibel <scheibel1@llnl.gov>

haampie committed:
docs search: rank api lowest and generated commands low (#47107)

haampie committed:
docs: tune ranking further (#47110)

promote hand-written docs, demote generated "docs" for sources, modules, packages.

haampie committed:
docs: do not promote build_systems/* at all (#47111)

alalazo committed:
Fix broken `spack find -u` (#47102)

fixes #47101

The bug was introduced in #33495, where `spack find was not updated,
and wasn't caught by unit tests.

Now a Database can accept a custom predicate to select the installation
records. A unit test is added to prevent regressions. The weird convention
of having `any` as a default value has been replaced by the more commonly
used `None`.

Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

scheibelp committed:
avoid double closing of fd in sub-processes (#47035)

Both `multiprocessing.connection.Connection.__del__` and `io.IOBase.__del__` called `os.close` on the same file descriptor. As of Python 3.13, this is an explicit warning. Ensure we close once by usef `os.fdopen(..., closefd=False)`

haampie committed:
buildcache: recognize . and .. as paths instead of names (#47105)

tgamblin committed:
mypy: work around typing issues with `functools.partial` (#47160)

haampie committed:
Replace MultiProcessFd with Connection objects

Connection objects are Python version, platform and multiprocessing
start method independent, so better to use those than a mix of plain
file descriptors and inadequate guesses in the child process whether it
was forked or not.

This also allows us to delete the now redundant MultiProcessFd class,
hopefully making things a bit easier to follow.

haampie committed:
Avoid a socket to communicate effectively a bit

haampie committed:
bootstrap: remove all system gnupg/patchelf executables (#47165)

haampie committed:
ensure write_fd.close() isn't called when sys.std* cannot be redirected

haampie committed:
compilers.yaml: require list of strings for modules (#47197)

haampie committed:
bootstrap: do not consider source when metadata file missing (#47278)

aphedges committed:
Fix typo in default `concretizer.yaml` (#47307)

This was caught by `codespell` when I copied the config file into an
internal repository.

aphedges committed:
Fix errors found by running `spack audit externals` (#47308)

The problem was that `+` is part of the regex grammar, so it needs to be
escaped.

aphedges committed:
Fix malformed RST link in documentation (#47309)

haampie committed:
autopush: run after install manifest is written

aphedges committed:
Remove trailing spaces in default YAML files (#47328)

caught by `prettier`

adamjstewart committed:
Docs: remove reference to pyspack (#47346)

haampie committed:
remove xfail marked hanging foreground/background tests

haampie committed:
database.py: remove process unsafe update_explicit (#47358)

Fixes an issue reported where `spack env depfile` + `make -j` would
non-deterministically refuse to mark all environment roots explicit.

`update_explicit` had the pattern

```python
rec = self._data[key]
with self.write_transaction():
    rec.explicit = explicit
```

but `write_transaction` may reinitialize `self._data`, meaning that
mutating `rec` won't mutate `self._data`, and the changes won't be
persisted.

Instead, use `mark` which has a correct implementation.

Also avoids the essentially incorrect early return in `update_explicit`
which is a pattern I don't think belongs in database.py: it branches on
possibly stale data to realize there is nothing to change, but in reality
it requires a write transaction to know that for a fact, but that would
defeat the purpose. So, leave this optimization to the call site.

alalazo committed:
Fix pickle round-trip of specs propagating variants (#47351)

This changes `Spec` serialization to include information about propagation for abstract specs.
This was previously not included in the JSON representation for abstract specs, and couldn't be
stored.

Now, there is a separate `propagate` dictionary alongside the `parameters` dictionary. This isn't
beautiful, but when we bump the spec version for Spack `v0.24`, we can clean up this and other
aspects of the schema.

alalazo commented:
Added #47351 to the list of things in the description that can be considered for backporting.

haampie committed:
packaging_guide.rst: explain forward and backward compat before the less common cases (#47402)

The idea is to go from most to least used: backward compat -> forward compat -> pinning on major or major.minor version -> pinning specific, concrete versions.

Further, the following

```python
   # backward compatibility with Python
   depends_on("python@3.8:")
   depends_on("python@3.9:", when="@1.2:")
   depends_on("python@3.10:", when="@1.4:")

   # forward compatibility with Python
   depends_on("python@:3.12", when="@:1.10")
   depends_on("python@:3.13", when="@:1.12")
   depends_on("python@:3.14")
```

is better than disjoint when ranges causing repetition of the rules on dependencies, and requiring frequent editing of existing lines after new releases are done:

```python
   depends_on("python@3.8:3.12", when="@:1.1")
   depends_on("python@3.9:3.12", when="@1.2:1.3")
   depends_on("python@3.10:3.12", when="@1.4:1.10")
   depends_on("python@3.10:3.13", when="@1.11:1.12")
   depends_on("python@3.10:3.14", when="@1.13:")

haampie committed:
libc.py: detect glibc also in chinese locale (#47434)

haampie committed:
getting_started.rst: fix list of spack deps (#47557)

haampie committed:
Set version to v0.22.3
